### ISSUE TITLE
Can't start node agent with Azure cloud-provider on OSE 3.7

### ISSUE THREAD

**Issue by edsonmarquezani** [openshift/origin#18115]
I'm deploying an Openshift Enterprise 3.7 on Azure with Ansible playbooks. Everything from RH repositories packages.

```
atomic-openshift-3.7.14-1.git.0.593a50e.el7.x86_64
atomic-openshift-clients-3.7.14-1.git.0.593a50e.el7.x86_64
atomic-openshift-docker-excluder-3.7.14-1.git.0.593a50e.el7.noarch
atomic-openshift-excluder-3.7.14-1.git.0.593a50e.el7.noarch
atomic-openshift-master-3.7.14-1.git.0.593a50e.el7.x86_64
atomic-openshift-node-3.7.14-1.git.0.593a50e.el7.x86_64
atomic-openshift-sdn-ovs-3.7.14-1.git.0.593a50e.el7.x86_64
tuned-profiles-atomic-openshift-node-3.7.14-1.git.0.593a50e.el7.x86_64
```

Ansible fails at `openshift_node : restart node`, because nodes can't get a subnet.
```
Jan 15 16:03:31 openshift-tst-master-1 atomic-openshift-node[122755]: F0115 16:03:31.681848  122755 network.go:45] SDN node startup failed: failed to get subnet for this host: openshift-tst-master-1, error: timed out waiting for the condition
```

After some investigation, I figured out the issue is somehow related to *cloud-provider* configuration, even though it seems to be correctly configured.
```
Jan 15 15:00:48 openshift-tst-master-1 atomic-openshift-node[83427]: I0115 15:00:48.734526   83427 azure.go:174] azure: using client_id+client_secret to retrieve access token
Jan 15 15:00:48 openshift-tst-master-1 atomic-openshift-node[83427]: I0115 15:00:48.734633   83427 node_config.go:120] Successfully initialized cloud provider: "azure" from the config file: "/etc/azure/azure.conf"
Jan 15 15:00:48 openshift-tst-master-1 atomic-openshift-node[83427]: I0115 15:00:48.933868   83427 server.go:652] cloud provider determined current node name to be openshift-tst-master-1
```

If I remove the *cloud-provider* configuration, it will successfully get the host subnet and the node will ingress in the cluster.
```
NAME                     HOST                     HOST IP         SUBNET           EGRESS IPS
[root@openshift-tst-master-1 ~]# oc get hostsubnets
openshift-tst-master-1   openshift-tst-master-1   10.200.47.153   10.254.16.0/23   []
```
```
[root@openshift-tst-master-1 ~]# oc get nodes
NAME                     STATUS    AGE       VERSION
openshift-tst-master-1   Ready     1m       v1.7.6+a08f5eeb62
```

Finally, if I put *cloud-provider* parameters back, and restart node agent, it'll keep working. This works for every node in my cluster.

`nodeName` is consistent between config file and the cloud.
```
[root@openshift-tst-master-1 ~]# grep nodeName /etc/origin/node/node-config.yaml
nodeName: openshift-tst-master-1
```

I could not find any other relevant logs on master and node components. Does anyone have a clue on this?

Ansible playbook variables are as follows:
```
openshift_cloudprovider_kind=azure
osm_controller_args={'cloud-provider': ['azure'], 'cloud-config': ['/etc/azure/azure.conf']}
osm_api_server_args={'cloud-provider': ['azure'], 'cloud-config': ['/etc/azure/azure.conf']}
openshift_node_kubelet_args={'cloud-provider': ['azure'], 'cloud-config': ['/etc/azure/azure.conf'], 'enable-controller-attach-detach': ['true']}
```

Obs: this same setup works on OSE 3.6.

**Comment by asatsi** [comment]
Was stuck with the issue and finally asked Microsoft support to help us out. I believe it was fixed by using this master: https://github.com/ivanthelad/openshift-origin/commits/master

**Comment by edsonmarquezani** [comment]
@asatsi Is it something related to configuration/setup or Openshift/Kubernetes implementation? I'm asking it because I don't use those default MS setup scripts. Instead, I have my own automation, and I'd like to know if I can fix it.

I noticed you now restart `openvswitch` before restarting nodes -> https://github.com/ivanthelad/openshift-origin/commit/2ca52f3e342faf24c9274efbbcfd6b5c2a1bf059

Is this it?

**Comment by asatsi** [comment]
@edsonmarquezani I have used the fix from the branch I mentioned which was done by Microsoft. Sorry don't have much details.

**Comment by edsonmarquezani** [comment]
As far as I could see, Microsoft scripts now sets the cluster up without any `cloud-provider` at first, doing that later, after nodes have ingressed in the cluster (as I have figured out on my tests). Seems like a workaround, actually, not a real fix for the problem (whose cause remains unknown for me).

**Comment by ivanthelad** [comment]
hey guys, 
specifically, whats required here is a restart of openshift multi-tenant sdn (same apples to the flat sdn )
https://github.com/ivanthelad/openshift-origin/blob/master/scripts/deployOpenShift.sh#L515
 
Executing `ansible all -b  -m service -a 'name=openvswitch state=restarted` and then the node service should fix it

**Comment by edsonmarquezani** [comment]
@ivanthelad It didn't work for me. I'm
- setting up a fresh cluster without cloud-provider
- re-configuring master components with cloud-provider Azure and restarting
- re-configuring node agents with cloud-provider Azure
- restarting `openvswitch`
- deleting node and restarting node agent

No luck so far.

**Comment by glennswest** [comment]
actually been injecting the azure.conf for quite a while in the ref arch. And I run it rather regularly, worked fine 3.5-3.6. Some time in december something changed, and no longer will the cluster work if you inject during the setup. The post cluster restart was not as reliable during this as the pre-install inject.

**Comment by edsonmarquezani** [comment]
Same behavior with Origin 3.7. SDN restart won't have any effect. Does anyone have a clue on this?

**Comment by edsonmarquezani** [comment]
Additional info. Debbuging node agent, I noticed the absence of these requests when it is Azure-enabled (request bodies ommited), which I suppose is the node registration. Can't tell why, though.
```
POST https://master.origin.naturacloud.com:8443/api/v1/nodes
  Request Headers:
  Accept: application/vnd.kubernetes.protobuf, */*
  Content-Type: application/vnd.kubernetes.protobuf
  User-Agent: openshift/v1.7.6+a08f5eeb62 (linux/amd64) kubernetes/c84beff
  Request Body: {"count":2,"lastTimestamp":"2018-01-23T19:24:09Z","message":"Node openshift-orig-infra-1 status is now: "Node openshift-orig-infra-1 status is now: NodeHasSufficientDisk"}

PATCH https://master.origin.naturacloud.com:8443/api/v1/namespaces/default/events/openshift-orig-infra-1.150c86dadfa7add7
  Request Headers:
  Content-Type: application/strategic-merge-patch+json
  Accept: application/vnd.kubernetes.protobuf, */*
  User-Agent: openshift/v1.7.6+a08f5eeb62 (linux/amd64) kubernetes/c84beff
```

So,

**Comment by edsonmarquezani** [comment]
Follow up.

Made it work **without removing nodes** (as recommended by Red Hat). So, the following steps **does work**.
- set up a fresh cluster without cloud-provider
- reconfigure master components with cloud-provider Azure and restart
- restart openvswitch
- reconfigure node agents with cloud-provider Azure and restart

(notice nodes wasn't deleted during the process)

Nonetheless, **nodes won't register when cloud-provider is enabled**.

**Comment by heatmiser** [comment]
@edsonmarquezani I experienced similar behavior to what you were seeing. I too am using automation outside of the MS arm template deployment, I'm utilizing Terraform to deploy everything.  In my particular case, I was setting the name of VM instances in Azure to a different value than the actual hostname configured within each VM.  When the cloud-provider configuration step ran and then nodes were removed from Kubernetes, the origin-node service would not start correctly.  Kubernetes could not register/re-add the nodes back, as they could not be found in Azure due to the Azure VM name/VM internal hostname mismatch.  Once I adjusted the Terraform template to use the same value for Azure VM name and internal VM hostname, everything deployed correctly to completion.  For the bash scripts that are utilized, I used my repo forked from the current master Microsoft/openshift-origin branch, with my only changes being the addition of support for padded numbering of VM hostnames.

**Comment by edsonmarquezani** [comment]
@heatmiser Thanks for you contribution, but that hasn't never been my case. I've been always using the default hostnames set by Azure, thus the same for internal hostname X VM name.
Anyway, I managed to workaround it, although the original problem remains (nodes won't register for the first time with cloud-provider set).

I also published my full automation, if anyone find it useful. https://github.com/edsonmarquezani/terraform-openshift-origin

**Comment by glennswest** [comment]
This seems to be a bit more complicated, in the timing of the setup and weather the registry is setup. In the case of single vm, its straight forward.

Form multi-vm:
Assuming:
  Setup the registry to use azure blob using the ansible variables. 
Then:
  Install the cluster (Without Cloud Provider)
  Setup the azure.conf on all nodes and restart the appropriate services
  Reset all the nodes openvswitch at the same time.

**Comment by mwitzenm** [comment]
@glennswest very interested in the outcome!! I have 2 different customers asking about this ( I believe this is how you were pulled into this at first ).

**Comment by maci0** [comment]
@edsonmarquezani do you still see this issue in 3.9 ?

**Comment by edsonmarquezani** [comment]
@maci0 I couldn't verify it yet. I may do it before the end of this month, though.

**Comment by edsonmarquezani** [comment]
@maci0 couldn't confirm if the issue is still there because of this other issue -> https://github.com/openshift/openshift-ansible/issues/7252 , which seems to occur only when I set cloud-provider on inventory.

I'm still setting cloud-provider after a "clean" setup (with success).

**Comment by maci0** [comment]
@edsonmarquezani thank you for the update, I appreciate it. 
Fortunately https://github.com/openshift/openshift-ansible/issues/7252 doesn't look like a big issue.

**Comment by jnach** [comment]
Wondering if there is a roadmap yet for azure integration in the inventory? We've solved with a post-install scripting but it adds a significant amount of time and won't be supported beyond the scope of the engagement.

**Comment by git001** [comment]
@maci0 
> do you still see this issue in 3.9 ?

Yes. we have seen the same behaviour in 3.9.30

**Comment by maci0** [comment]
This issue was greatly improved with a latter errata release of 3.9.
It seems to be fixed in 3.10 entirely

**Comment by git001** [comment]
> This issue was greatly improved with a later errata release of 3.9.

We have tried the following versions 3.9.43 (latest) and 3.9.41 in both versions is the node in NotReady State and don't recover by there own.

We have to  reboot it several times until they come ready.
